tests/cli/test_datasets.py

```diff
@@ -1147,3 +1147,17 @@ def test_dataset_rm_tags_failure(tmpdir, runner, project, client):
         catch_exceptions=False,
     )
     assert 2 == result.exit_code
+
+
+def test_dataset_clean_up_when_add_fails(runner, client):
+    """Test project is cleaned when dataset add fails for a new dataset."""
+    # add a non-existing path to a new dataset
+    result = runner.invoke(
+        cli,
+        ['dataset', 'add', 'new-dataset', 'non-existing-file'],
+        catch_exceptions=True,
+    )
+
+    assert result.exit_code == 1
+    assert not (client.renku_path / 'refs' / 'datasets' /
```

Comment:
wouldn't `exists` be enough here?

Reply:
The test wasn't failing with `exists`. It came out that for symlinks `exists` checks if the target exists or not:
```
Note If the path points to a symlink, exists() returns whether the symlink points to an existing file or directory.
```

Reply:
but shouldn't the symlink be removed?

Reply:
It is removed with this fix. I could not write a failing test for this using `exists`. When using `exists` the test will always pass either because the symlink is removed (this fix) or because the symlink is there but its target is not there (without this fix).

I can see that this might be a bit confusing, I'll change the logic to:
```
assert not ref.is_symlink() and not ref.exists()
```